### DISCUSSION TITLE
[apex] Apex code analysis

### DISCUSSION THREAD

**Discussion by Puneet1726** [pmd/pmd#3029] [Q&A] [answered]
Hi,

Currently I am running PMD analysis though command line. Is there a way to segregate issues like Critical, Bug, Warning and Bloker? 

I am exporting the analysis on a html file. 

pmd.bat -d "path" -f html -R "path\Apex Rules.xml" -reportfile "path\result.html".

Can we integrate PMD with any other tool like Sonar or codeacy to get the details on the UI?

Thanks,
Puneet

**Comment by renatoliveira** [comment]
> Is there a way to segregate issues like Critical, Bug, Warning and Bloker?

It's just my guessing but I think there isn't. There's always an option to write a script to read the output and do what you want, though.

> Can we integrate PMD with any other tool like Sonar or codeacy to get the details on the UI?

You can check for tools that support the outputs that PMD produces. 

I don't know other tools, but I'm pretty sure you can integrate it with GitLab. [GitLab can parse the JUnit result format](https://docs.gitlab.com/ee/ci/junit_test_reports.html#how-it-works).

**Comment by rsoesemann** [comment, marked as answer]
Please watch my free Pluralsight course on how to integrate using Codacy: https://www.pluralsight.com/courses/play-by-play-automated-code-analysis-in-salesforce. Sonar currently doesn’t support apex to the full extend. I cannot recommend it.

**Comment by adangel** [comment]
You can use the priority of the rules. You of course need to decide on your own, which rule you consider yourself as Blocker, Critical, Warning, Bug.

Follow the documentation here: https://pmd.github.io/latest/pmd_userdocs_configuring_rules.html#message-and-priority-overriding in order to configure the priority for your rules.

Then you might want to try out different report formats: https://pmd.github.io/latest/pmd_userdocs_cli_reference.html#available-report-formats

You could also just use xml + xslt to render a report.

There are many integrations of PMD into various tools, see https://pmd.github.io/latest/pmd_userdocs_tools_ci.html and https://pmd.github.io/latest/pmd_userdocs_tools.html

**Comment by Puneet1726** [comment]
Thanks @adangel .

One final question, can we add priority details in html report?
Currently it is not showing, can we add priority after Problem?
# | File | Line | Problem
-- | -- | -- | --


Thanks,
Puneet
